--- a/sources/drivers/rtl8188eus/hal/rtl8188e/usb/usb_halinit.c
+++ b/sources/drivers/rtl8188eus/hal/rtl8188e/usb/usb_halinit.c
@@ -281,9 +281,9 @@
 #ifdef CONFIG_SUPPORT_USB_INT
 	/* REG_USB_SPECIAL_OPTION - BIT(4) */
 	/* 0; Use interrupt endpoint to upload interrupt pkt */
 	/* 1; Use bulk endpoint to upload interrupt pkt,	 */
 	usb_opt = rtw_read8(Adapter, REG_USB_SPECIAL_OPTION);
 
-	if ((IS_FULL_SPEED_USB(Adapter))
+	if (IS_FULL_SPEED_USB(Adapter)
 #ifdef CONFIG_USB_INTERRUPT_IN_PIPE
 	    || (pdev->RtInPipe[REALTEK_USB_IN_INT_EP_IDX] == 0x05)
 #endif
 	   )
