diff --git a/kernel/runtime/ksud_integration.c b/kernel/runtime/ksud_integration.c
index ba7d3b95..45d9a2f5 100644
--- a/kernel/runtime/ksud_integration.c
+++ b/kernel/runtime/ksud_integration.c
@@ -165,6 +165,7 @@ void stop_execve_hook();
 void ksu_stop_input_hook_runtime();
 
 #ifdef KSU_KPROBES_HOOK
+bool ksu_init_rc_hook __read_mostly = true;
 static struct work_struct __maybe_unused stop_init_rc_hook_work;
 static struct work_struct __maybe_unused stop_execve_hook_work;
 static struct work_struct __maybe_unused stop_input_hook_work;
@@ -889,6 +890,26 @@ void ksu_wait_stop_input_hook(void)
 {
 	wait_for_completion(&stop_input_hook_work_complete);
 }
+
+int __maybe_unused ksu_handle_vfs_read(struct file **file_ptr, char __user **buf_ptr,
+					 size_t *count_ptr, loff_t **pos)
+{
+	return 0;
+}
+
+void __maybe_unused ksu_handle_newfstat_ret(unsigned int *fd,
+					 struct stat __user **statbuf_ptr)
+{
+	return;
+}
+
+#if defined(__ARCH_WANT_STAT64) || defined(__ARCH_WANT_COMPAT_STAT64)
+void __maybe_unused ksu_handle_fstat64_ret(unsigned long *fd,
+					 struct stat64 __user **statbuf_ptr)
+{
+	return;
+}
+#endif
 #else
 static int ksu_execve_ksud_common(const char __user *filename_user,
 				  struct user_arg_ptr *argv)
@@ -1038,8 +1059,9 @@ void ksu_handle_fstat64_ret(unsigned long *fd, struct stat64 __user **statbuf_pt
 void stop_init_rc_hook()
 {
 #ifdef KSU_KPROBES_HOOK
+	ksu_init_rc_hook = false;
 	bool ret = schedule_work(&stop_init_rc_hook_work);
-    pr_info("unregister init_rc_hook kprobe: %d!\n", ret);
+	pr_info("unregister init_rc_hook kprobe: %d!\n", ret);
 #else
 	ksu_init_rc_hook = false;
 	pr_info("stop init_rc_hook\n");
